### ISSUE TITLE
Changed workspaces filter should account for lockfile updates

### ISSUE THREAD

**Issue by finn-orsini** [vercel/turbo#2162]
### TL;DR 

Turbo [`--filter` by changed workspaces](https://turborepo.org/docs/core-concepts/filtering#filter-by-changed-workspaces) should mark workspaces as changed if they have been impacted by a lockfile update. 

### Full Description 

Currently,  when filtering by changed workspaces, lockfile updates do not mark workspaces as changed. Example: 

- `git clone git@github.com:finn-orsini/turbo-lockfile-mvr.git` 
- `cd turbo-lockfile-mvr` 
- `git checkout changed-lockfile && yarn`  ([branch diff](https://github.com/finn-orsini/turbo-lockfile-mvr/compare/main...changed-lockfile))
- `yarn build-filter-dry` 

The output of the above is: 

```
$ turbo run build --dry=json --filter="[origin/main]"
{
"packages": [
  "//"
],
"tasks": []
}
```

The [`//` token](https://turborepo.org/docs/core-concepts/filtering#the-workspace-root) indicates that something in the root of the workspace changed, but does not tell us if this is a meaningful / build impacting change (i.e. a lockfile update), or a root change that will not impact any builds (a README update).

Lockfile updates directly impact application builds. Given a build system that relies on `--filter` to detect which applications need to build & deploy, breaking lockfile changes can currently be merged to `main` without proper validation. They will be discovered the next time an impacted app is changed, which results in developer confusion: Why are my changes breaking in this way, which seems unrelated to the changes on my branch? How did `main` get into this state?

#### Notes

- This is distinct from existing issues on if caching behavior should/shouldn't include the lockfile, and from adding the lockfile to [`globalDependencies`](https://turborepo.org/docs/reference/configuration#globaldependencies) or [`inputs`](https://turborepo.org/docs/reference/configuration#inputs) in turbo.json. In this issue, we are only concerned with the `--filter` logic, not if the resulting tasks are cached or not.
- The example repo is contrived, but we saw this in our prod repo with a lockfile-only upgrade of our design system (using `yarn upgrade-interactive`) not marking any workspaces as changed when all should have been.

### Describe the solution you'd like

Turbo filtering by changed workspaces should include workspaces impacted by a lockfile update. 

Ideally the solution would be "smart" - that is, not all lockfile changes result in all workspaces building, only those directly impacted by the update. 

This could potentially be implemented by using the same logic as the `prune` option, by comparing the minimal lockfile on each ref (i.e. origin/main and the current commit) for each workspace.

We expect that this behavior would be opt-in, at least to start.

### Describe alternatives you've considered

Our current alternative is a naive solution - whenever the lockfile changes, we omit the `--filter` flag to build all workspaces. This could be built into turbo via a `globalFilterDependencies` option in `turbo.json`. 

This method is not ideal. Instead of false negatives where apps are not being built that should be, we now have false positives where apps are built that didn't need to be.

**Comment by gsoltis** [comment]
I think this would be a great feature to have. 

From an initial glance, some of the pieces we'd need to implement are:
 - use `git` to get the lockfile at the `fromRef` [here](https://github.com/vercel/turborepo/blob/main/cli/internal/scope/scope.go#L147)
 - parse the lockfile, similar to how we do it [here](https://github.com/vercel/turborepo/blob/main/cli/internal/context/context.go#L134)
 - iterate the list of packages, and for each one check for equality on the resolved [Subgraph](https://github.com/vercel/turborepo/blob/main/cli/internal/lockfile/lockfile.go#L17)
 
Some pieces to resolve:
 - Do we need the old `package.json` files to properly prune the old lockfile? Or maybe we can consider any failure to get a subtree as a change? We may need to make the subgraph resolution more robust to handle these cases.

**Comment by nathanhammond** [comment]
I _think_ the current `package.json` resolutions should still work because in a lockfile-only update the specifiers should stay the same? @chris-olszewski would be the person who can confirm that hypothesis.

**Comment by nathanhammond** [comment]
One more note: I believe that we should treat this as a _bug_.

The root lockfile updated `jest`. The UI package didn't change but the resolution for its jest specifiers _did_ change.

In other words, following the setup in @finn-orsini's report and then running `turbo run build --dry=json --filter="web...[origin/main]"` should return `"packages": ["//", "web"]`. It's a correctness issue that it does not.

/cc @tknickman

**Comment by tknickman** [comment]
https://github.com/vercel/turborepo/pull/2211 implements the naive approach (for now) and will return all workspaces when changes are detected to:

1. turbo.json
2. root package.json
3. package manager lock file

**Comment by attila** [comment]
Yes, that change in #2211 now triggers a full build on all our 100+ workspaces every time anything changes in package-lock.json in the root, regardless of relevance. The naive approach could use some refinement 😅

**Comment by yarinsa** [comment]
Hey having the same issue and figured it could be a nice contribution to turbo. 
[See my issue](https://github.com/vercel/turbo/issues/2606) which was resolved by #2211 . 

I think that the expected logic should be:
1. Collect all yarn lock changes since commit X
2. Filter all the dependencies that their    final resolutions  **changed** comparing to the yarn lock before commit X
3. Check all the packages in the workspace that include those dependencies
4. Return number 3 as the output of filter

Questions:
1. What do you guys think about this? does this make sense? am I missing something?
2. Anyone who already contributed are familiar with the code that can direct me to the right area? 

@tknickman Maybe I can take it with you as a side community contribution?

**Comment by tknickman** [comment]
Yea that sounds good, @chris-olszewski tagging you in here too for the lockfile portion

**Comment by attila** [comment]
@yarinsa it looks great but ideally this logic should be supported on all lockfiles, not only yarn. We for example just use npm workspaces so package-lock.json support would be desirable here.

**Comment by yarinsa** [comment]
> @yarinsa it looks great but ideally this logic should be supported on all lockfiles, not only yarn. We for example just use npm workspaces so package-lock.json support would be desirable here.

Finally got to work on that logic last week, but I spent a day trying to setup the env. If I can find help with that I assume implementing this for all lock won't be much different

**Comment by chris-olszewski** [comment]
Hi all, I was able to set aside some time last week to put together #3250 which should be a complete implementation. I still need to do quite a bit of testing before I'll feel comfortable shipping it, but I think most of the heavy lifting is done.

**Comment by attila** [comment]
Yes it's been working great for us, a life saver of not to rebuild everything every time on any lockfile change anymore. It was a real pain on our 140+ workspace npm repo, but now all is well 👏 

Thank you for fixing this!
